### ISSUE TITLE
`onBeforeRender` in `.page.ts` in SPA/Client Routing Mode causes server-side render

### ISSUE THREAD

**Issue by alexkreidler** [brillout/vite-plugin-ssr#204]
I changed my default server rendering code, `_default.page.server.tsx` (see below) to return an empty HTML document, and for the client to simply call `ReactDOM.render` on load. This SPA mode helped me avoid a rabbit hole of bugs with Vite's SSR system `ssrModuleLoader` with some particular packages that I was importing.

However, I wanted to add some `pageContext` on the frontend to change the layout of the `PageShell` that I'm using. This is my page:

```tsx
import React from "react";
import { VisualizePage } from "../features/visualization/components/VisualizePage";

export { Page };

function Page() {
  return <VisualizePage />;
}

export function onBeforeRender() {
  return {
    pageContext: {
      pageProps: {
        noPadding: true,
      }
    }
  }
}
```

Unfortunately, this led to the Vite SSR errors listed below, and it seemed Vite was running its module analysis on frontend React dependencies on the server side. This was perplexing to me, because **my default server renderer does not render any react components**. I have a hunch that somehow `vite-plugin-ssr` is interpreting that because I have that function there, it should replace my default server code with some code that renders the components server side and passes data to the client. But I'm not sure; I really don't know enough about the project internals so I'd really appreciate any guidance you can provide.

**Expected behavior:** do not do a server-side render and only call `onBeforeRender` on the client.

When I hit reload in my browser or access the page for the first time, it displays a 500 Internal Server Error message. But if I make any change to the `visualize.page.tsx` file, like adding or removing spaces/newlines, and then saving, that causes a hot module reload, which causes the components to render in the client without a problem. This was confusing for me for a bit as I though I had somehow fixed it by tweaking the file, but the issue remains.

Let me know if I can provide more information. Thanks for this great library!

<details>
<summary>Errors</summary>

```/home/alex/c2/loqu4/node_modules/.pnpm/compassql@0.21.2_vega@5.21.0/node_modules/compassql/build/src/query/encoding.js:1
import { isObject } from 'datalib/src/util';
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at nodeRequire (/home/alex/c2/loqu4/node_modules/.pnpm/vite@2.6.7/node_modules/vite/src/node/ssr/ssrModuleLoader.ts:183:15)
    at ssrImport (/home/alex/c2/loqu4/node_modules/.pnpm/vite@2.6.7/node_modules/vite/src/node/ssr/ssrModuleLoader.ts:94:14)
    at eval (/models/shelf/spec/index.ts:1:37)
10:51:23 PM [vite] Error when evaluating SSR module /models/shelf/index.ts:
/home/alex/c2/loqu4/node_modules/.pnpm/compassql@0.21.2_vega@5.21.0/node_modules/compassql/build/src/query/encoding.js:1
import { isObject } from 'datalib/src/util';
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at nodeRequire (/home/alex/c2/loqu4/node_modules/.pnpm/vite@2.6.7/node_modules/vite/src/node/ssr/ssrModuleLoader.ts:183:15)
    at ssrImport (/home/alex/c2/loqu4/node_modules/.pnpm/vite@2.6.7/node_modules/vite/src/node/ssr/ssrModuleLoader.ts:94:14)
    at eval (/models/shelf/spec/index.ts:1:37)
10:51:23 PM [vite] Error when evaluating SSR module /selectors/shelf.ts:
/home/alex/c2/loqu4/node_modules/.pnpm/compassql@0.21.2_vega@5.21.0/node_modules/compassql/build/src/query/encoding.js:1
import { isObject } from 'datalib/src/util';
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at nodeRequire (/home/alex/c2/loqu4/node_modules/.pnpm/vite@2.6.7/node_modules/vite/src/node/ssr/ssrModuleLoader.ts:183:15)
    at ssrImport (/home/alex/c2/loqu4/node_modules/.pnpm/vite@2.6.7/node_modules/vite/src/node/ssr/ssrModuleLoader.ts:94:14)
    at eval (/models/shelf/spec/index.ts:1:37)
10:51:23 PM [vite] Error when evaluating SSR module /selectors/index.ts:
/home/alex/c2/loqu4/node_modules/.pnpm/compassql@0.21.2_vega@5.21.0/node_modules/compassql/build/src/query/encoding.js:1
import { isObject } from 'datalib/src/util';
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at nodeRequire (/home/alex/c2/loqu4/node_modules/.pnpm/vite@2.6.7/node_modules/vite/src/node/ssr/ssrModuleLoader.ts:183:15)
    at ssrImport (/home/alex/c2/loqu4/node_modules/.pnpm/vite@2.6.7/node_modules/vite/src/node/ssr/ssrModuleLoader.ts:94:14)
    at eval (/models/shelf/spec/index.ts:1:37)
10:51:23 PM [vite] Error when evaluating SSR module /features/visualization/components/MainView.tsx:
/home/alex/c2/loqu4/node_modules/.pnpm/compassql@0.21.2_vega@5.21.0/node_modules/compassql/build/src/query/encoding.js:1
import { isObject } from 'datalib/src/util';
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at nodeRequire (/home/alex/c2/loqu4/node_modules/.pnpm/vite@2.6.7/node_modules/vite/src/node/ssr/ssrModuleLoader.ts:183:15)
    at ssrImport (/home/alex/c2/loqu4/node_modules/.pnpm/vite@2.6.7/node_modules/vite/src/node/ssr/ssrModuleLoader.ts:94:14)
    at eval (/models/shelf/spec/index.ts:1:37)
10:51:23 PM [vite] Error when evaluating SSR module /features/visualization/components/VisualizePage.tsx:
/home/alex/c2/loqu4/node_modules/.pnpm/compassql@0.21.2_vega@5.21.0/node_modules/compassql/build/src/query/encoding.js:1
import { isObject } from 'datalib/src/util';
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at nodeRequire (/home/alex/c2/loqu4/node_modules/.pnpm/vite@2.6.7/node_modules/vite/src/node/ssr/ssrModuleLoader.ts:183:15)
    at ssrImport (/home/alex/c2/loqu4/node_modules/.pnpm/vite@2.6.7/node_modules/vite/src/node/ssr/ssrModuleLoader.ts:94:14)
    at eval (/models/shelf/spec/index.ts:1:37)
10:51:23 PM [vite] Error when evaluating SSR module /pages/visualize.page.tsx:
/home/alex/c2/loqu4/node_modules/.pnpm/compassql@0.21.2_vega@5.21.0/node_modules/compassql/build/src/query/encoding.js:1
import { isObject } from 'datalib/src/util';
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at nodeRequire (/home/alex/c2/loqu4/node_modules/.pnpm/vite@2.6.7/node_modules/vite/src/node/ssr/ssrModuleLoader.ts:183:15)
    at ssrImport (/home/alex/c2/loqu4/node_modules/.pnpm/vite@2.6.7/node_modules/vite/src/node/ssr/ssrModuleLoader.ts:94:14)
    at eval (/models/shelf/spec/index.ts:1:37)

```

</details>

<details>
<summary>_default.page.server.tsx</summary>

```tsx
import { escapeInject, dangerouslySkipEscape } from "vite-plugin-ssr";
import type { PageContext } from "./types";
import type { PageContextBuiltIn } from "vite-plugin-ssr";

export const passToClient = ["pageProps", "urlPathname"];
import { makeGoogleFontsURL } from "@/utils/fonts";

function render(pageContext: PageContextBuiltIn & PageContext) {
  const { documentProps } = pageContext;
  const title = (documentProps && documentProps.title) || "mytitle";
  const desc =
    (documentProps && documentProps.description) ||
    "mydesc";

  const documentHtml = escapeInject`<!DOCTYPE html>
    <html lang="en">
      <head>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <link rel="preconnect" href="https://fonts.gstatic.com" />
        <link
          href="${makeGoogleFontsURL(["Inter"])}"
          rel="stylesheet"
        />
        <meta name="description" content="${desc}" />
        <title>${title}</title>
      </head>
      <body>
        <div id="page-view"></div>
      </body>
    </html>`;

  return {
    documentHtml,
    pageContext: {},
  };
}

export { render };
```

</summary>

**Comment by brillout** [comment]
You're right, after some thoughts, I consider this a `vite-plugin-ssr` design flaw. I'll fix this ASAP.

The solution I've in mind so far is to allow you to define `Page` and `onBeforeRender()` in `.page.client.ts`, which means that they will be only loaded in the browser and never in Node.js. (This introduces breaking changes so I'm going to thoroughly think it through.)

**Comment by brillout** [comment]
(SPA use case has high prio.)

**Comment by alexkreidler** [comment]
Great, I really appreciate the focus on SPAs. Your plan sounds good-- never want to rush breaking changes! 

In the meantime I tried to remove every last bit of pageContext from my app, thought I fixed it (because I was doing a lot of testing just with the frontend routing via `useClientRouter`) but then started adding pageContext back in the form of URL params etc and it reappeared.

I also found that `pageContext.isHydration` in `render()` of `useClientRouter()` was set to true even though I "disabled SSR" as described above, so it's probably related.

Another somewhat unrelated question is about exporting to HTML. I have this SPA and even though it has JavaScript and is mostly rendered through React, I want to distribute it as static HTML and JS files that I can put on a CDN. 

Is there a way to do this with `vite-plugin-ssr`? I know the equivalent with NextJS is `next export`. The default `vite --ssr` output requires that Node server.
I looked at and tried the `vite-plugin-ssr prerender` CLI command but unfortunately got the same server-side Vite SSR errors. (BTW for anyone curious I raised an issue for them in https://github.com/vitejs/vite/issues/5673)
Basically, I want to skip the Vite SSR related things and just output an HTML file that has a link to JS module (probably not even a bundle via rollup b/c that also involves vite stuff) that can be served without a NodeJS server.
If it doesn't exist yet, I think it would be a great new feature and could potentially attract NextJS people with Vite's speed and the flexibility of using different view frameworks, but I don't know how difficult an implementation would be. (I can make a new issue for this as well)

Thanks for looking into this!

**Comment by brillout** [comment]
In the meantime what you can do

```js
// some-page.page.js

export { Page }
export { onBeforeRender }

async function Page() {
  if (isNodejs()) {
    return;
  }
  const { Page } = await import('./some-page.page-client-only.js')
  return Page.apply(null, ...arguments)
}

async function onBeforeRender() {
  if (isNodejs()) {
    return;
  }
  const { onBeforeRender } = await import('./some-page.page-client-only.js')
  return onBeforeRender.apply(null, ...arguments)
}

function isNodejs() {
  return typeof window === "undefined"
}
```

Let me know if this temporary hack works for you. In principle it should.

**Comment by brillout** [comment]
Yes, see https://vite-plugin-ssr.com/pre-rendering

And yes it works with both the new design and the temporary hack.

**Comment by alexkreidler** [comment]
I don't want any actual React rendering (SSR) in this case, just a bit of transpiling/bundling of my source files into output JS and then creating an HTML file that links to the bundle. But I guess maybe the issue is that to do any type of that transformation of source files, Vite has to be involved and so I get the those errors. Here's an example:

```
> vite-plugin-ssr prerender --partial

vite-plugin-ssr 0.3.27 pre-rendering HTML...
/home/alex/c2/loqu4/node_modules/.pnpm/compassql@0.21.2_vega@5.21.0/node_modules/compassql/build/src/query/encoding.js:1
import { isObject } from 'datalib/src/util';
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Module.<anonymous> (/home/alex/c2/loqu4/dist/server/assets/visualize.page.ce9b2f41.js:44:16)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
 ELIFECYCLE  Command failed with exit code 1.
```

So I should probably figure out how to fix those errors with Vite upstream or think about moving to another build tool.

**Comment by alexkreidler** [comment]
I may try your workaround tomorrow. Thanks again!

**Comment by brillout** [comment]
Is `compassql` supposed to run in Node.js or in the browser?

The workaround should fix your issues because it essentially removes all your browser-side code from being loaded in Node.js.

**Comment by alexkreidler** [comment]
It can run on both (it's related to the data visualization libraries Vega and Vega-Lite), although I admit its a finnicky package otherwise I wouldnt be getting so many issues.

I understand that the workaround may stop that issue but I've run into the general issue where i'm getting Vite errors several times in the past few hours after I removed pagecontext. So it may have nothing to do with the actual React rendering, maybe some type of static-ish analysis.

Also I have read those docs on pre-rendering. (and almost all the docs)

Also, for me, the main issue is not that I need to customize pagecontext but to avoid or fix vite ssr errors. 
So changing my code to stop customizing page context was much easier than changing a lot more code to fix the import errors at least for now.

The errors have occurred somewhat consistently with Redux although it may be a coincidence as i have much of the imports of the data vis libraries there.

Also apologies if this is a bit incoherent- i'm up late.

**Comment by brillout** [comment]
So far I see two problems:
 1 The design flaw I mentioned
 2. `compassql` importing problems

Let me know if you are confident there is anything else. (About Redux: my guess is that using Redux forced you to move code to the server leading to hit 2. again.)

The temporary workaround should fix both 1. and 2.

Keep me updated how the workaround works out for you.

**Comment by brillout** [comment]
I'm making progress on resolving the design flaw.

**Comment by brillout** [comment]
Alright, the new design is now clear in my head.

@alexkreidler Let me know how urgent this is for you, and I'll prioritize the implemention accordingly.

**Comment by alexkreidler** [comment]
This isn't super urgent for me as I don't have to use `onBeforeRender` right now. Thanks!

**Comment by brillout** [comment]
I will implement this after Telefunc's beta release.

**Comment by mkhuda** [comment]
I just noticed some libraries that require access to `window` objects, such as `react-leaflet` will throws error. Is there a good way to use `clientOnly` libs?

**Comment by brillout** [comment]
@mkhuda Is using a dynamic `import()` to load react-leaflet an option for you?

**Comment by mkhuda** [comment]
@brillout the workaround I'm implementing now is by using `React.lazy` and `React.Suspense` on clientSide only.
Here's an example:
```jsx
import React from 'react';

const LeafletMapComponent = React.lazy(() => import('@components/Map'));

export { Page };

function Page() {
  const isSSR = typeof window === 'undefined';

  return (
    <>
      {!isSSR && (
        <React.Suspense fallback={<div />}>
          <LeafletMapComponent />
        </React.Suspense>
      )}
    </>
  );
}
```
I think this is the cleanest way for my case. But if you have another solution, I'll try. :grin:

**Comment by brillout** [comment]
@mkhuda That's actually what I had in mind.

**Comment by brillout** [comment]
@mkhuda Added new docs for this: https://vite-plugin-ssr.com/dynamic-import. See the included React example, which I believe is a slight improvement compared to what you have.

```jsx
// pages/pick-location.page.jsx

import React from 'react'
import { Loading } from './Loading'

export function Page() {
  // Users will see the text `Please pick a location` before the browser loads the `<Map>`
  // component's code. (This is the case anyways when using SSR, but `import()` allows the
  // user to not only see but also interact with our page before `<Map>` is loaded.)
  return <>
    Please pick a location.
    <Map />
  </>
}

function Map() {
  const isBrowser = typeof window !== 'undefined'
  const isNodejs = !isBrowser

  // We skip rendering the interactive map on the server-side (SSR'ing a map is mostly useless).
  if (isNodejs) {
    return <Loading />
  }

  // We lazily load `<SomeHeavyMapComponent>`
  const SomeHeavyMapComponent = React.lazy(() => import('some-heavy-map-component'))
  return (
    <React.Suspense fallback={<Loading />}>
      <SomeHeavyMapComponent />
    </React.Suspense>
  )
}
```

**Comment by mkhuda** [comment]
@brillout Thank you very much, that is cool and pretty straightforward. I am currently planning to migrate some projects using `vite-plugin-ssr`.

**Comment by brillout** [comment]
👍 Let me know if you have questions.

**Comment by brillout** [comment]
Fixed in `0.4`.
